features/config/dao/api/src/main/java/org/opennms/features/config/dao/api/ConfigSchema.java

```diff
@@ -0,0 +1,111 @@
+/*******************************************************************************
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2019-2019 The OpenNMS Group, Inc.
+ * OpenNMS(R) is Copyright (C) 1999-2019 The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is a registered trademark of The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+package org.opennms.features.config.dao.api;
+
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import javax.persistence.Transient;
+import java.io.Serializable;
+import java.lang.reflect.Constructor;
+
+/**
+ * This class include meta-data of the config
+ */
+public class ConfigSchema<T extends XmlConfigConverter> implements Serializable {
```

Comment:
I don't see any schema besides xsd. The idea is to make it as generic as possible, may be one day we want other format. I will change the name to ConfigConverter make it less confuse.

Reply:
XSDs is what we have for schemas today, but we do have parts of the configuration that are not currently covered by any schema i.e. .properties files and .cfg files (for OSGi Blueprints). We'll want some way to manage these with CM as well.

One way would be to craft XSDs, but this is going to be cumbersome, particularly when it comes to expressing validation rules. Given that we don't really use the full XSD, and only map it to a simplified schema, we could come up with a more concise way of expressing these.

It's still relativalye early in the design, but just wanted to make sure that we keep this in mind.

Reply:
Piggybacking on Jesse's comment, there are _already_ validation issues with some of our configs that just plain can't be handled by XSDs because they're based on more complex logic, so having an official API way to do additional validation in code is a good idea either way.

Reply:
I think we might end up having two layers of validation:
1. A generic validation that can be derived from an XSD or some other means (e.g. what we prototyped in liquibase). The validation will live inside the ConfigurationManager and be called at every change to the configuration.
2. An optional complex validation that can be added when registering a configuration. This would be a java class provided by the service which registers the configuration. This callback will be called at every change to the configuration. This way the ConfigurationManager can reject a modified configuration before it is stored.